kani-compiler/src/codegen_cprover_gotoc/codegen/operand.rs

```diff
@@ -304,24 +304,58 @@ impl<'tcx> GotocCtx<'tcx> {
+                    let overall_type = self.codegen_ty(ty);
+                    // If there is no field, then it's just a ZST
                     if variant.fields.is_empty() {
                         if adt.is_struct() {
-                            let overall_t = self.codegen_ty(ty);
-                            Expr::struct_expr_from_values(overall_t, vec![], &self.symbol_table)
+                            Expr::struct_expr_from_values(overall_type, vec![], &self.symbol_table)
                         } else {
                             unimplemented!()
                         }
                     } else {
-                        // otherwise, there is just one field, which is stored as the scalar data
-                        let field = &variant.fields[0usize.into()];
-                        let fty = field.ty(self.tcx, subst);
-
-                        let overall_t = self.codegen_ty(ty);
+                        // Otherwise, there is at least one field. Again, there are two cases.
+                        let overall_type = self.codegen_ty(ty);
                         if adt.is_struct() {
-                            self.codegen_single_variant_single_field(s, span, overall_t, fty)
+                            if variant.fields.len() == 1 {
+                                // There is a single field which we associate with the scalar data.
+                                let field = variant.single_field();
+                                let field_type = field.ty(self.tcx, subst);
+                                self.codegen_single_variant_single_field(
+                                    s,
+                                    span,
+                                    overall_type,
+                                    field_type,
+                                )
+                            } else {
+                                // There are multiple fields, but only one is related to the scalar data.
+                                // The rest of them correspond to phantom data (ZST).
+                                let field_types: Vec<Ty<'_>> =
+                                    variant.fields.iter().map(|f| f.ty(self.tcx, subst)).collect();
+                                // Check that there is a single non-phantom field
+                                let non_phantom_field_types: Vec<_> =
+                                    field_types.iter().filter(|t| !t.is_phantom_data()).collect();
+                                assert!(
+                                    non_phantom_field_types.len() == 1,
+                                    "error: expected exactly one field that is not phantom data"
+                                );
+                                let field_values: Vec<Expr> = field_types
+                                    .iter()
+                                    .map(|t| {
+                                        if t.is_phantom_data() {
```

Comment:
I don't think this will in fact fix the issue I just pasted above. Can you please check?

I recommend you use the following function instead:

https://github.com/model-checking/kani/blob/dba16744ee1d9299b379d9f876c28a8d07376933/kani-compiler/src/codegen_cprover_gotoc/codegen/typ.rs#L456-L458

Reply:
Thanks! You're right, other ZST would cause the compiler to crash (for example, the test you sent in the other comment).